### ISSUE TITLE
Variables used only by directives should be removed from the request

### ISSUE THREAD

**Issue by alexlafroscia** [HoudiniGraphql/houdini#308]
I am running into an issue while using the `@append(parentID: ID)` directive as a hint to Houdini about how to correctly insert a newly-created record into a list

I have a query like this in my application, which is displayed on a route like `/list/:primaryKey`:

```gql
query ListPage($primaryKey: Int!) {
  list(primaryKey: $primaryKey) {
    id # `ID` for the purposes of implementing the `Node` interface
    primaryKey # Actual `id` in the database; queries use this data instead

    items @list(name: "ListPageItems") {
      id
      content
    }
}
```

Within that page, I have a button to create a new item within that list:

```gql
mutation AddItemToList($listPrimaryKey: Int!) {
  createItemInList(listPrimaryKey: $listPrimaryKey) {
    ...ListPageItems_insert
  }
}
```

This works fine on the first list page navigated to (say, `/list/1`) but breaks if I perform a navigation to another list (like `/list/2`) without fully reloading the page. If I stay on the initial list, the list is inserted into automatically. If I navigate to the second list and try to perform the mutation, the operation is successful but the item just doesn't appear in _either_ of the lists. Refreshing `/list/2` then shows the item correctly, since it's loaded anew from the API.

In an attempt to fix this, I tried to use the `@append` directive, updating my mutation like so:

```gql
mutation AddItemToList($listID: ID!, $listPrimaryKey: Int!) {
  createItemInList(listPrimaryKey: $listPrimaryKey) {
    ...ListPageItems_insert @append(parentID: $listID)
  }
}
```

However, this causes me to run into a new error: the server complains that `$listID` is never actually used in the mutation, and thus should not be provided.

This isn't wrong, from the perspective of the API, since the entire `@append` directive only exists on the client; as far as I can tell, the actual API call has the directive removed. However, I can't just remove the variable from the mutation either, since the client needs it.

I _think_ that any variables used only by a directive, and never used in the actual GraphQL query, should be stripped out of the query before sending it to the API. Does that sound like the right solution to this problem?

**Comment by AlecAivazis** [comment]
Oh good point! Thanks for such a detailed description of the problem. I think your proposed solution is right. Implementing this will be slightly tricky but if you want to give it a shot, I'd be happy to provide any guidance you need. Otherwise, I'll try to get something out relatively soon since you're currently blocked

**Comment by alexlafroscia** [comment]
I'll see if I can put some time toward fixing this myself after work today!

>  since you're currently blocked

No worries here; the app I'm working on isn't mission-critical and shipping it with a know bug like this isn't the end of the world.

**Comment by AlecAivazis** [comment]
Awesome! If you want a more direct way of getting ahold of me feel free to reach out on discord

**Comment by AlecAivazis** [comment]
Thinking more closely about your problem and I don't think that `@parentID` will help in your situation. I'm guessing you wouldn't be able to publish your work stuff somewhere I can check out but if you had the time to put together a repo that shows the issue, I would gladly take a look and can figure out what's going on

**Comment by AlecAivazis** [comment]
this is now fixed in version `0.14.8` but i'm still not sure it will help through your original issue. If you wouldn't mind creating a separate issue with these details we can try to figure it out together

**Comment by alexlafroscia** [comment]
> but i'm still not sure it will help through your original issue. If you wouldn't mind creating a separate issue with these details we can try to figure it out together

Unfortunately, that seems to be the case!

I'll open up a new issue with the relevant information from this one. I think I'm going to need a whole demo app to really display the bug, which... might take some time to put together!
